src/Bicep.Core/Registry/OciModuleRegistry.cs

```diff
@@ -76,6 +76,41 @@ public override bool IsModuleRestoreRequired(OciArtifactModuleReference referenc
                 !this.FileResolver.FileExists(this.GetModuleFileUri(reference, ModuleFileType.Metadata));
         }
 
+        public override async Task<bool> CheckModuleExists(OciArtifactModuleReference reference)
+        {
+            try
+            {
+                // Get module
+                await this.client.PullArtifactAsync(configuration, reference);
+            }
+            catch (RequestFailedException exception) when (exception.Status == 404)
+            {
+                // Found module but tag doesn't exist
+                return false;
+            }
+            catch (ExternalModuleException exception) when
+                (exception.InnerException is RequestFailedException &&
+                ((RequestFailedException)exception.InnerException).Status == 404)
+            {
+                // Found no module at all
+                return false;
+            }
+            catch (RequestFailedException exception) when (exception.Message.StartsWith("The requested digest does not match the digest of the received manifest."))
```

Comment:
>The requested digest does not match the digest of the received manifest. [](http://example.com/codeflow?start=89&length=72)

That's not the right exception message to catch the wrong artifact type. In the case where the OCI artifact has the wrong artifactType or has the wrong layer media type, we would throw in one of these places:
* [https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L175](https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L175)
* [https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L212](https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L212)

In addition we shouldn't assert any specific exception message since we control both the catching and the throwing of the exception. My suggestion would be to add a `Code` or `Kind` enum property to the `InvalidModuleException` class with these values: `NotSpecified`, `WrongArtifactType`, and `WrongModuleLayerMediaType` and then update the throw sites to set it. (In places not relevant to these codes, can just default to `NotSpecified`.)

(Btw, Digest mismatch on the manifest would be a bug or compromise of the registry and we throw a different exception for this at [https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L165](https://github.com/Azure/bicep/blob/a1d217e3bb3d049fc5c6c21b466c3a29585a6c69/src/Bicep.Core/Registry/AzureContainerRegistryManager.cs#L165))

Reply:
The reason for matching that message was that I got that message back when trying to overwrite a artifact of a completely different type, in this case I was trying with the `hello-world:latest` container from Docker to see what would happen, and that was the result. Not sure if that's a case intended to be caught in the mismatch error you referred.

Reply:
Disregarding the case with the `hello-world` image, I added a "Kind" enum according to your suggestion and made `InvalidModuleException` public in its own file so that we can catch it and check the `Kind` property for `WrongArtifactType` or `WrongModuleLayerMediaType`.

Reply:
> The reason for matching that message was that I got that message back when trying to overwrite a artifact of a completely different type, in this case I was trying with the `hello-world:latest` container from Docker to see what would happen, and that was the result. It seems that the blob client itself actually verifies that the digest matches. Not sure if that's a case intended to be caught in the mismatch error you refer to.

Ahh got it. Didn't realize that the client was already doing those checks. They are the same case, so we might be doing extra work.

Fix looks good to me also.